src/operator/distinct.rs

```diff
@@ -281,177 +268,248 @@ where
+                    output.push((Z::item_from(key.clone(), val.clone()), output_weight));
+                }
 
-            // Update output.
-            if delta_old != delta_new {
-                builder.push((Z::item_from(value.clone(), ()), delta_new + delta_old.neg()));
+                if let Some(t) = time_of_interest {
+                    self.keys_of_interest
+                        .entry(t)
+                        .or_insert_with(Vec::new)
+                        .push(((key.clone(), val.clone()), Present));
+                }
             }
+        }
+    }
 
-            // Record next_ts in `self.future_updates`.
-            if let Some(next_ts) = next_ts {
-                let idx: usize = next_ts.inner() as usize;
-                self.future_updates
-                    .entry(idx as u32)
-                    .or_insert_with(BTreeSet::new)
-                    .insert(value.clone());
-            }
-        } else if weight.ge0() && !weight.is_zero() {
-            builder.push((Z::item_from(value.clone(), ()), HasOne::one()));
+    fn init_distinct_vals(vals: &mut [(Option<T::Time>, Z::R)], ts: Option<T::Time>) {
+        if vals.len() == 1 {
+            vals[0] = (ts, HasZero::zero());
+        } else {
+            let half_len = vals.len() >> 1;
+            Self::init_distinct_vals(
+                &mut vals[0..half_len],
+                ts.as_ref()
+                    .and_then(|ts| ts.checked_recede(half_len.ilog2() as Scope)),
+            );
+            Self::init_distinct_vals(&mut vals[half_len..], ts);
+        }
+    }
+
+    fn clear_distinct_vals(&mut self) {
+        for (_ts, val) in self.distinct_vals.iter_mut() {
+            *val = HasZero::zero();
+        }
+    }
+
+    /// Compute partial derivative.
+    fn partial_derivative(vals: &[(Option<T::Time>, Z::R)]) -> Z::R {
+        // Split vals in two halves.  Compute `partial_derivative` recursively
+        // for each half, return the difference.
+        if vals.len() == 1 {
+            vals[0].1.clone()
+        } else {
+            Self::partial_derivative(&vals[vals.len() >> 1..])
+                + Self::partial_derivative(&vals[0..vals.len() >> 1]).neg()
         }
     }
 }
 
-impl<Z, T> Operator for DistinctTrace<Z, T>
+impl<Z, T> Operator for DistinctIncremental<Z, T>
 where
-    Z: ZSet,
-    T: BatchReader<Key = Z::Key, Val = (), Time = NestedTimestamp32, R = Z::R>,
+    Z: IndexedZSet,
+    T: BatchReader<Key = Z::Key, Val = Z::Val, R = Z::R>,
 {
     fn name(&self) -> Cow<'static, str> {
-        Cow::Borrowed("DistinctTrace")
+        Cow::Borrowed("DistinctIncremental")
     }
 
     fn metadata(&self, meta: &mut OperatorMeta) {
-        let size: usize = self.future_updates.values().map(BTreeSet::len).sum();
-        let bytes = self.future_updates.size_of();
+        let size: usize = self.keys_of_interest.values().map(Vec::len).sum();
+        let bytes = self.keys_of_interest.size_of();
```

Comment:
```suggestion
        let size: usize = self.keys_of_interest.values().map(Vec::len).sum();
        let bytes = size_of::size_of_values([
            &self.keys_of_interes as &dyn SizeOf,
            &self.distinct_vals as &dyn SizeOf,
        ]);
```
I'm assuming from the comment that `distinct_vals` is always kept empty within `self` but if not then its length should be added to `size`

Reply:
Now that I look at it more, I'd almost prefer if we just put a `#[derive(SizeOf)]` onto the `DistinctIncremental` struct itself and then just call `self.size_of()`